File: android/src/main/kotlin/com/example/seasonal_app_icon/SeasonalAppIconPlugin.kt
package com.example.seasonal_app_icon

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** SeasonalAppIconPlugin */
class SeasonalAppIconPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware {

    private lateinit var channel: MethodChannel
    private var context: Context? = null
    private var activity: Activity? = null
    private var currentIconName: String? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "seasonal_app_icon")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "setIcon" -> handleSetIcon(call, result)
            "getCurrentIcon" -> handleGetCurrentIcon(result)
            "supportsAlternateIcons" -> handleSupportsAlternateIcons(result)
            "getAvailableIcons" -> handleGetAvailableIcons(result)
            else -> result.notImplemented()
        }
    }

    private fun handleSetIcon(call: MethodCall, result: Result) {
        val iconName = call.argument<String?>("iconName")
        val ctx = context ?: run {
            result.success(mapOf("success" to false, "error" to "Context not available"))
            return
        }

        try {
            val packageManager = ctx.packageManager
            val packageName = ctx.packageName

            // Get all activity-aliases from the manifest
            val availableIcons = getActivityAliases(ctx)

            // Disable all activity-aliases first
            availableIcons.forEach { alias ->
                val componentName = ComponentName(packageName, "$packageName.$alias")
                packageManager.setComponentEnabledSetting(
                    componentName,
                    PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                    PackageManager.DONT_KILL_APP
                )
            }

            // Enable the main activity or the selected alias
            if (iconName == null) {
                // Reset to default - enable the main activity
                val mainActivityName = getMainActivityName(ctx)
                if (mainActivityName != null) {
                    val mainComponent = ComponentName(packageName, mainActivityName)
                    packageManager.setComponentEnabledSetting(
                        mainComponent,
                        PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                        PackageManager.DONT_KILL_APP
                    )
                }
                currentIconName = null
            } else {
                // Disable main activity and enable the selected alias
                val mainActivityName = getMainActivityName(ctx)
                if (mainActivityName != null) {
                    val mainComponent = ComponentName(packageName, mainActivityName)
                    packageManager.setComponentEnabledSetting(
                        mainComponent,
                        PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                        PackageManager.DONT_KILL_APP
                    )
                }

                val aliasComponent = ComponentName(packageName, "$packageName.$iconName")
                packageManager.setComponentEnabledSetting(
                    aliasComponent,
                    PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                    PackageManager.DONT_KILL_APP
                )
                currentIconName = iconName
            }

            result.success(mapOf("success" to true, "iconName" to iconName))
        } catch (e: Exception) {
            result.success(mapOf("success" to false, "error" to (e.message ?: "Unknown error")))
        }
    }

    private fun handleGetCurrentIcon(result: Result) {
        result.success(currentIconName)
    }

    private fun handleSupportsAlternateIcons(result: Result) {
        // Android supports alternate icons via activity-alias
        result.success(true)
    }

    private fun handleGetAvailableIcons(result: Result) {
        val ctx = context ?: run {
            result.success(listOf<String>())
            return
        }
        result.success(getActivityAliases(ctx))
    }

    private fun getActivityAliases(ctx: Context): List<String> {
        val aliases = mutableListOf<String>()
        try {
            val packageInfo = ctx.packageManager.getPackageInfo(
                ctx.packageName,
                PackageManager.GET_ACTIVITIES or PackageManager.GET_DISABLED_COMPONENTS
            )
            packageInfo.activities?.forEach { activityInfo ->
                val name = activityInfo.name
                // Activity aliases typically have a pattern like "packageName.AliasName"
                if (name.contains(".") && !name.endsWith(".MainActivity")) {
                    val aliasName = name.substringAfterLast(".")
                    if (aliasName != "MainActivity") {
                        aliases.add(aliasName)
                    }
                }
            }
        } catch (e: Exception) {
            // Ignore errors
        }
        return aliases
    }

    private fun getMainActivityName(ctx: Context): String? {
        try {
            val packageInfo = ctx.packageManager.getPackageInfo(
                ctx.packageName,
                PackageManager.GET_ACTIVITIES
            )
            packageInfo.activities?.forEach { activityInfo ->
                if (activityInfo.name.endsWith(".MainActivity")) {
                    return activityInfo.name
                }
            }
        } catch (e: Exception) {
            // Ignore errors
        }
        return null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: ios/Classes/SeasonalAppIconPlugin.swift
import Flutter
import UIKit

public class SeasonalAppIconPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "seasonal_app_icon", binaryMessenger: registrar.messenger())
    let instance = SeasonalAppIconPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setIcon":
      handleSetIcon(call: call, result: result)
    case "getCurrentIcon":
      handleGetCurrentIcon(result: result)
    case "supportsAlternateIcons":
      handleSupportsAlternateIcons(result: result)
    case "getAvailableIcons":
      handleGetAvailableIcons(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleSetIcon(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard UIApplication.shared.supportsAlternateIcons else {
      result(["success": false, "error": "Alternate icons are not supported on this device"])
      return
    }

    guard let args = call.arguments as? [String: Any] else {
      result(["success": false, "error": "Invalid arguments"])
      return
    }

    let iconName = args["iconName"] as? String

    UIApplication.shared.setAlternateIconName(iconName) { error in
      if let error = error {
        result(["success": false, "error": error.localizedDescription])
      } else {
        result(["success": true, "iconName": iconName as Any])
      }
    }
  }

  private func handleGetCurrentIcon(result: @escaping FlutterResult) {
    if UIApplication.shared.supportsAlternateIcons {
      result(UIApplication.shared.alternateIconName)
    } else {
      result(nil)
    }
  }

  private func handleSupportsAlternateIcons(result: @escaping FlutterResult) {
    result(UIApplication.shared.supportsAlternateIcons)
  }

  private func handleGetAvailableIcons(result: @escaping FlutterResult) {
    guard let icons = Bundle.main.object(forInfoDictionaryKey: "CFBundleIcons") as? [String: Any],
          let alternateIcons = icons["CFBundleAlternateIcons"] as? [String: Any] else {
      result([String]())
      return
    }
    result(Array(alternateIcons.keys))
  }
}
